pkg/graph/builder.go

```diff
@@ -682,117 +683,87 @@ func extractDependencies(env *cel.Env, expression string, resourceNames []string
 // we evaluate B's CEL expressions against 2 emulated resources A and C, and so
 // on.
 func validateResourceCELExpressions(resources map[string]*Resource, instance *Resource) error {
-	resourceNames := maps.Keys(resources)
-	// We also want to allow users to refer to the instance spec in their expressions.
-	resourceNames = append(resourceNames, "schema")
+	resourceNames := []string{"schema"}
 	conditionFieldNames := []string{"schema"}
 
-	env, err := krocel.DefaultEnvironment(krocel.WithResourceIDs(resourceNames))
-	if err != nil {
-		return fmt.Errorf("failed to create CEL environment: %w", err)
-	}
-	instanceEmulatedCopy := instance.emulatedObject.DeepCopy()
-	if instanceEmulatedCopy != nil && instanceEmulatedCopy.Object != nil {
-		delete(instanceEmulatedCopy.Object, "apiVersion")
-		delete(instanceEmulatedCopy.Object, "kind")
-		delete(instanceEmulatedCopy.Object, "status")
-	}
-
 	for _, resource := range resources {
-		for _, resourceVariable := range resource.variables {
-			for _, expression := range resourceVariable.Expressions {
-				err := validateCELExpressionContext(env, expression, resourceNames)
-				if err != nil {
-					return fmt.Errorf("failed to validate expression context: '%s' %w", expression, err)
-				}
+		// First validate includeWhen expressions
+		for _, includeWhenExpression := range resource.includeWhenExpressions {
+			instanceEnv, err := krocel.DefaultEnvironment(krocel.WithResourceIDs(resourceNames))
+			if err != nil {
+				return fmt.Errorf("failed to create CEL environment: %w", err)
+			}
 
-				// create context
-				context := map[string]*Resource{}
-				for resourceName, contextResource := range resources {
-					// exclude the resource we are validating
-					if resourceName != resource.id {
-						context[resourceName] = contextResource
-					}
-				}
-				// add instance spec to the context
-				context["schema"] = &Resource{
-					emulatedObject: &unstructured.Unstructured{
-						Object: instanceEmulatedCopy.Object,
-					},
-				}
+			err = validateCELExpressionContext(instanceEnv, includeWhenExpression, conditionFieldNames)
+			if err != nil {
+				return fmt.Errorf("failed to validate expression context: '%s' %w", includeWhenExpression, err)
+			}
 
-				_, err = dryRunExpression(env, expression, context)
-				if err != nil {
-					return fmt.Errorf("failed to dry-run expression %s: %w", expression, err)
-				}
+			// Type check the includeWhen expression
+			ast, iss := instanceEnv.Compile(includeWhenExpression)
+			if iss != nil && iss.Err() != nil {
+				return fmt.Errorf("failed to compile includeWhen expression %s: %w", includeWhenExpression, iss.Err())
```

Comment:
Compiling a CEL expression doesn't necessarily mean that you're type checking it

Reply:
The change in error message format should be consistent. We should keep the error message format consistent throughout the codebase.